Feature request: Assert.Multiple() could return an IDisposable, avoiding passing an Action around.

Would it be possible to be able to write code like:
```
// Act
var result = GetResult();

// Assert
using (Assert.Multiple())
{
    Assert.That(result.A, Is.EqualTo("A"));
    Assert.That(result.B, Is.EqualTo("B"));
}
```

This would hopefully avoid capturing unnecessary context into the Action (because it wouldn't exist).  Async code becomes slightly simpler (no need to specify `async` again) also.

It might also make things like #4242 easier to get right.

Considerations:
- As written, this would not conflict, but we would certainly want to drive users towards this as a preferred option over the existing `Assert.Multiple(() => { ... });` option.  So we could consider using a different name to avoid ambiguity.

@stevenaw I tried to convert the `AssertionScope` into a `ref struct`, but that makes `async` methods impossible:

![image](https://github.com/user-attachments/assets/90ab0c8d-a08f-4c48-aa71-b3a3e2e8fa21)

Even when compiling with SDK 9.0-preview6, I still get an error:

```
Assert.cs(244,20): error CS4007: Instance of type 'NUnit.Framework.Assert.Scope' cannot be preserved across 'await' or 'yield' boundary.
```

Even if it worked, we would then requires user to use C#13 which is not yet released and in a non-LTS SDK.

So I reverted back to the class.

--- a/src/NUnitFramework/framework/Assert.cs
+++ b/src/NUnitFramework/framework/Assert.cs
@@ -263,6 +263,7 @@ namespace NUnit.Framework
         {
             private readonly TestExecutionContext _context;
             private readonly int _assertionCountWhenEnteringScope;
+            private readonly int _multipleAssertLevelInScope;
 
             public AssertionScope()
             {
@@ -270,11 +271,17 @@ namespace NUnit.Framework
                 Guard.OperationValid(_context is not null, "There is no current test execution context.");
 
                 _assertionCountWhenEnteringScope = _context.CurrentResult.AssertionResults.Count;
-                _context.MultipleAssertLevel++;
+                _multipleAssertLevelInScope = ++_context.MultipleAssertLevel;
             }
 
             public void Dispose()
             {
+                if (TestExecutionContext.CurrentContext != _context ||
+                    _context.MultipleAssertLevel != _multipleAssertLevelInScope)
+                {
+                    throw new InvalidOperationException("The assertion scope was disposed out of order.");
+                }
+
                 _context.MultipleAssertLevel--;
 
                 if (_context is { MultipleAssertLevel: 0, CurrentResult: { PendingFailures: > 0 } })

--- a/src/NUnitFramework/framework/Assert.cs
+++ b/src/NUnitFramework/framework/Assert.cs
@@ -4,6 +4,7 @@ using System;
 using System.ComponentModel;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.Threading;
 using System.Threading.Tasks;
 using NUnit.Framework.Interfaces;
 using NUnit.Framework.Internal;
@@ -265,6 +266,8 @@ namespace NUnit.Framework
             private readonly int _assertionCountWhenEnteringScope;
             private readonly int _multipleAssertLevelInScope;
 
+            private int _isDisposed;
+
             public AssertionScope()
             {
                 _context = TestExecutionContext.CurrentContext;
@@ -276,6 +279,9 @@ namespace NUnit.Framework
 
             public void Dispose()
             {
+                if (Interlocked.Exchange(ref _isDisposed, 1) == 1)
+                    return; // Already disposed.
+
                 if (TestExecutionContext.CurrentContext != _context ||
                     _context.MultipleAssertLevel != _multipleAssertLevelInScope)
                 {

--- a/src/NUnitFramework/framework/Assert.cs
+++ b/src/NUnitFramework/framework/Assert.cs
@@ -213,28 +213,9 @@ namespace NUnit.Framework
         /// <param name="testDelegate">A TestDelegate to be executed in Multiple Assertion mode.</param>
         public static void Multiple(TestDelegate testDelegate)
         {
-            TestExecutionContext context = TestExecutionContext.CurrentContext;
-            Guard.OperationValid(context is not null, "There is no current test execution context.");
-
-            var oldCount = context.CurrentResult.AssertionResults.Count;
-            context.MultipleAssertLevel++;
-
-            try
+            using (EnterMultipleScope())
             {
                 testDelegate();
-            }
-            finally
-            {
-                context.MultipleAssertLevel--;
-            }
-
-            if (context is { MultipleAssertLevel: 0, CurrentResult: { PendingFailures: > 0 } })
-            {
-                context.CurrentResult.RecordTestCompletion();
-                if (context.CurrentResult.AssertionResults.Count > oldCount)
-                {
-                    throw new MultipleAssertException(context.CurrentResult);
-                }
             }
         }
 
@@ -246,28 +227,9 @@ namespace NUnit.Framework
         /// <param name="testDelegate">A TestDelegate to be executed in Multiple Assertion mode.</param>
         public static void Multiple(AsyncTestDelegate testDelegate)
         {
-            TestExecutionContext context = TestExecutionContext.CurrentContext;
-            Guard.OperationValid(context is not null, "There is no current test execution context.");
-
-            var oldCount = context.CurrentResult.AssertionResults.Count;
-            context.MultipleAssertLevel++;
-
-            try
+            using (EnterMultipleScope())
             {
                 AsyncToSyncAdapter.Await(testDelegate.Invoke);
-            }
-            finally
-            {
-                context.MultipleAssertLevel--;
-            }
-
-            if (context is { MultipleAssertLevel: 0, CurrentResult: { PendingFailures: > 0 } })
-            {
-                context.CurrentResult.RecordTestCompletion();
-                if (context.CurrentResult.AssertionResults.Count > oldCount)
-                {
-                    throw new MultipleAssertException(context.CurrentResult);
-                }
             }
         }
 
@@ -279,24 +241,50 @@ namespace NUnit.Framework
         /// <param name="testDelegate">An AsyncTestDelegate to be executed in Multiple Assertion mode.</param>
         public static async Task MultipleAsync(AsyncTestDelegate testDelegate)
         {
-            TestExecutionContext context = TestExecutionContext.CurrentContext;
-            Guard.OperationValid(context is not null, "There is no current test execution context.");
-
-            context.MultipleAssertLevel++;
-
-            try
+            using (EnterMultipleScope())
             {
                 await testDelegate();
             }
-            finally
+        }
+
+        /// <summary>
+        /// Enters a multiple assert scope.
+        /// Wraps code containing a series of assertions, which should all
+        /// be executed, even if they fail. Failed results are saved and
+        /// reported when the returned IDisposable is disposed.
+        /// </summary>
+        /// <returns>An <see cref="IDisposable"/> which when disposed leaves the multiple assertion scope.</returns>
+        public static IDisposable EnterMultipleScope()
+        {
+            return new AssertionScope();
+        }
+
+        private sealed class AssertionScope : IDisposable
+        {
+            private readonly TestExecutionContext _context;
+            private readonly int _assertionCountWhenEnteringScope;
+
+            public AssertionScope()
             {
-                context.MultipleAssertLevel--;
+                _context = TestExecutionContext.CurrentContext;
+                Guard.OperationValid(_context is not null, "There is no current test execution context.");
+
+                _assertionCountWhenEnteringScope = _context.CurrentResult.AssertionResults.Count;
+                _context.MultipleAssertLevel++;
             }
 
-            if (context is { MultipleAssertLevel: 0, CurrentResult: { PendingFailures: > 0 } })
+            public void Dispose()
             {
-                context.CurrentResult.RecordTestCompletion();
-                throw new MultipleAssertException(context.CurrentResult);
+                _context.MultipleAssertLevel--;
+
+                if (_context is { MultipleAssertLevel: 0, CurrentResult: { PendingFailures: > 0 } })
+                {
+                    _context.CurrentResult.RecordTestCompletion();
+                    if (_context.CurrentResult.AssertionResults.Count > _assertionCountWhenEnteringScope)
+                    {
+                        throw new MultipleAssertException(_context.CurrentResult);
+                    }
+                }
             }
         }
 

--- a/src/NUnitFramework/framework/Internal/Commands/TestMethodCommand.cs
+++ b/src/NUnitFramework/framework/Internal/Commands/TestMethodCommand.cs
@@ -47,7 +47,10 @@ namespace NUnit.Framework.Internal.Commands
             if (_testMethod.HasExpectedResult)
                 Assert.That(result, Is.EqualTo(_testMethod.ExpectedResult));
 
-            context.CurrentResult.SetResult(ResultState.Success);
+            if (context.MultipleAssertLevel > 0)
+                context.CurrentResult.SetResult(ResultState.Error, $"Test completed with {context.MultipleAssertLevel} active assertion scopes.");
+            else
+                context.CurrentResult.SetResult(ResultState.Success);
 
             if (context.CurrentResult.AssertionResults.Count > 0)
                 context.CurrentResult.RecordTestCompletion();

--- a/src/NUnitFramework/testdata/AssertMultipleData.cs
+++ b/src/NUnitFramework/testdata/AssertMultipleData.cs
@@ -525,6 +525,15 @@ namespace NUnit.TestData.AssertMultipleData
             outerScope.Dispose();
             innerScope.Dispose();
         }
+
+        [Test]
+        public void ScopeReleasedTwice()
+        {
+            using IDisposable scope = Assert.EnterMultipleScope();
+            Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+            Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+            scope.Dispose();
+        }
     }
 
     internal class ComplexNumber

--- a/src/NUnitFramework/testdata/AssertMultipleData.cs
+++ b/src/NUnitFramework/testdata/AssertMultipleData.cs
@@ -492,6 +492,39 @@ namespace NUnit.TestData.AssertMultipleData
                 }
             }
         }
+
+        [Test]
+        public void NonReleasedScope()
+        {
+            Assert.EnterMultipleScope();
+            Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+            Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+        }
+
+        [Test]
+        public void NonReleasedScopes()
+        {
+            Assert.EnterMultipleScope();
+            Assert.That(2 + 2, Is.EqualTo(4));
+
+            Assert.EnterMultipleScope();
+            Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+            Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+        }
+
+        [Test]
+        public void ScopeReleasedOutOfOrder()
+        {
+            IDisposable outerScope = Assert.EnterMultipleScope();
+            Assert.That(2 + 2, Is.EqualTo(4));
+
+            IDisposable innerScope = Assert.EnterMultipleScope();
+            Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+            Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+
+            outerScope.Dispose();
+            innerScope.Dispose();
+        }
     }
 
     internal class ComplexNumber

--- a/src/NUnitFramework/testdata/AssertMultipleData.cs
+++ b/src/NUnitFramework/testdata/AssertMultipleData.cs
@@ -305,6 +305,17 @@ namespace NUnit.TestData.AssertMultipleData
         }
 
         [Test]
+        public void ExceptionThrownAfterTwoFailures_EnterScope()
+        {
+            using (Assert.EnterMultipleScope())
+            {
+                Assert.That(2 + 2, Is.EqualTo(5), "Failure 1");
+                Assert.That(1, Is.EqualTo(0), "Failure 2");
+                throw new Exception("Simulated Error");
+            }
+        }
+
+        [Test]
         public void ExceptionThrownAfterWarning()
         {
             Assert.Multiple(() =>
@@ -363,6 +374,18 @@ namespace NUnit.TestData.AssertMultipleData
         }
 
         [Test]
+        public async Task ThreeAssertsSucceed_Async_EnterScope()
+        {
+            using (Assert.EnterMultipleScope())
+            {
+                await Task.Delay(100);
+                Assert.That(2 + 2, Is.EqualTo(4));
+                Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+                Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+            }
+        }
+
+        [Test]
         public void NestedBlock_ThreeAssertsSucceed_Async()
         {
             Assert.Multiple(async () =>
@@ -400,6 +423,26 @@ namespace NUnit.TestData.AssertMultipleData
         }
 
         [Test]
+        public async Task TwoNestedBlocks_ThreeAssertsSucceed_Async_EnterScope()
+        {
+            using (Assert.EnterMultipleScope())
+            {
+                using (Assert.EnterMultipleScope())
+                {
+                    await Task.Delay(100);
+                    Assert.That(2 + 2, Is.EqualTo(4));
+                }
+
+                using (Assert.EnterMultipleScope())
+                {
+                    Assert.That(Complex.RealPart, Is.EqualTo(5.2));
+                    await Task.Delay(100);
+                    Assert.That(Complex.ImaginaryPart, Is.EqualTo(3.9));
+                }
+            }
+        }
+
+        [Test]
         public void TwoAsserts_BothAssertsFail_Async()
         {
             Assert.Multiple(async () =>
@@ -428,6 +471,26 @@ namespace NUnit.TestData.AssertMultipleData
                     Assert.That(Complex.ImaginaryPart, Is.EqualTo(4.2), "ImaginaryPart");
                 });
             });
+        }
+
+        [Test]
+        public async Task TwoNestedBlocks_TwoAssertsFail_Async_EnterScope()
+        {
+            using (Assert.EnterMultipleScope())
+            {
+                using (Assert.EnterMultipleScope())
+                {
+                    await Task.Delay(100);
+                    Assert.That(2 + 2, Is.EqualTo(5));
+                }
+
+                using (Assert.EnterMultipleScope())
+                {
+                    await Task.Delay(100);
+                    Assert.That(Complex.RealPart, Is.EqualTo(5.2), "RealPart");
+                    Assert.That(Complex.ImaginaryPart, Is.EqualTo(4.2), "ImaginaryPart");
+                }
+            }
         }
     }
 

--- a/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
+++ b/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
@@ -47,6 +47,8 @@ namespace NUnit.Framework.Tests.Assertions
         [TestCase(nameof(AM.TwoAsserts_BothAssertsFail_Async), 2, "RealPart", "ImaginaryPart")]
         [TestCase(nameof(AM.TwoNestedBlocks_TwoAssertsFail_Async), 3, "Expected: 5", "ImaginaryPart")]
         [TestCase(nameof(AM.TwoNestedBlocks_TwoAssertsFail_Async_EnterScope), 3, "Expected: 5", "ImaginaryPart")]
+        [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures), 2, "Failure 1", "Failure 2")]
+        [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures_EnterScope), 2, "Failure 1", "Failure 2")]
         public void AssertMultipleFails(string methodName, int asserts, params string[] assertionMessages)
         {
             CheckResult(methodName, ResultState.Failure, asserts, assertionMessages);
@@ -61,8 +63,6 @@ namespace NUnit.Framework.Tests.Assertions
         }
 
         [TestCase(nameof(AM.ExceptionThrown), 0)]
-        [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures), 2, "Failure 1", "Failure 2", "Simulated Error")]
-        [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures_EnterScope), 2, "Failure 1", "Failure 2", "Simulated Error")]
         [TestCase(nameof(AM.ExceptionThrownAfterWarning), 0, "WARNING", "Simulated Error")]
         public void AssertMultipleErrorTests(string methodName, int asserts, params string[] assertionMessages)
         {

--- a/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
+++ b/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
@@ -80,6 +80,15 @@ namespace NUnit.Framework.Tests.Assertions
             Assert.That(result.Message, Contains.Substring($"{invalidAssert} may not be used in a multiple assertion block."));
         }
 
+        [TestCase(nameof(AM.NonReleasedScope), 2, "Test completed with 1 active assertion scopes")]
+        [TestCase(nameof(AM.NonReleasedScopes), 3, "Test completed with 2 active assertion scopes")]
+        [TestCase(nameof(AM.ScopeReleasedOutOfOrder), 3, "The assertion scope was disposed out of order")]
+        public void NonOrOutOfOrderReleaseScope(string methodName, int asserts, string errorMessage)
+        {
+            ITestResult result = CheckResult(methodName, ResultState.Error, asserts);
+            Assert.That(result.Message, Contains.Substring(errorMessage));
+        }
+
         [Test]
         public async Task AssertMultipleAsyncSucceeds()
         {
@@ -104,8 +113,8 @@ namespace NUnit.Framework.Tests.Assertions
 
             PlatformInconsistency.MonoMethodInfoInvokeLosesStackTrace.SkipOnAffectedPlatform(() =>
             {
-                if (result.ResultState.Status == TestStatus.Failed)
-                    Assert.That(result.StackTrace, Is.Not.Null.And.Contains(methodName), "StackTrace");
+                if (result.ResultState.Status == TestStatus.Failed && result.StackTrace is not null)
+                    Assert.That(result.StackTrace, Does.Contain(methodName), "StackTrace");
             });
 
             if (result.AssertionResults.Count > 0)

--- a/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
+++ b/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
@@ -28,6 +28,7 @@ namespace NUnit.Framework.Tests.Assertions
         [TestCase(nameof(AM.NestedBlock_ThreeAssertsSucceed_Async), 3)]
         [TestCase(nameof(AM.TwoNestedBlocks_ThreeAssertsSucceed_Async), 3)]
         [TestCase(nameof(AM.TwoNestedBlocks_ThreeAssertsSucceed_Async_EnterScope), 3)]
+        [TestCase(nameof(AM.ScopeReleasedTwice), 2)]
         public void AssertMultipleSucceeds(string methodName, int asserts)
         {
             CheckResult(methodName, ResultState.Success, asserts);

--- a/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
+++ b/src/NUnitFramework/tests/Assertions/AssertMultipleTests.cs
@@ -24,8 +24,10 @@ namespace NUnit.Framework.Tests.Assertions
         [TestCase(nameof(AM.ThreeWarnIf_AllPass), 3)]
         [TestCase(nameof(AM.ThreeWarnUnless_AllPass), 3)]
         [TestCase(nameof(AM.ThreeAssertsSucceed_Async), 3)]
+        [TestCase(nameof(AM.ThreeAssertsSucceed_Async_EnterScope), 3)]
         [TestCase(nameof(AM.NestedBlock_ThreeAssertsSucceed_Async), 3)]
         [TestCase(nameof(AM.TwoNestedBlocks_ThreeAssertsSucceed_Async), 3)]
+        [TestCase(nameof(AM.TwoNestedBlocks_ThreeAssertsSucceed_Async_EnterScope), 3)]
         public void AssertMultipleSucceeds(string methodName, int asserts)
         {
             CheckResult(methodName, ResultState.Success, asserts);
@@ -44,6 +46,7 @@ namespace NUnit.Framework.Tests.Assertions
         [TestCase(nameof(AM.WarningAfterTwoAssertsFail), 2, "Expected: 5", "ImaginaryPart", "WARNING")]
         [TestCase(nameof(AM.TwoAsserts_BothAssertsFail_Async), 2, "RealPart", "ImaginaryPart")]
         [TestCase(nameof(AM.TwoNestedBlocks_TwoAssertsFail_Async), 3, "Expected: 5", "ImaginaryPart")]
+        [TestCase(nameof(AM.TwoNestedBlocks_TwoAssertsFail_Async_EnterScope), 3, "Expected: 5", "ImaginaryPart")]
         public void AssertMultipleFails(string methodName, int asserts, params string[] assertionMessages)
         {
             CheckResult(methodName, ResultState.Failure, asserts, assertionMessages);
@@ -59,6 +62,7 @@ namespace NUnit.Framework.Tests.Assertions
 
         [TestCase(nameof(AM.ExceptionThrown), 0)]
         [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures), 2, "Failure 1", "Failure 2", "Simulated Error")]
+        [TestCase(nameof(AM.ExceptionThrownAfterTwoFailures_EnterScope), 2, "Failure 1", "Failure 2", "Simulated Error")]
         [TestCase(nameof(AM.ExceptionThrownAfterWarning), 0, "WARNING", "Simulated Error")]
         public void AssertMultipleErrorTests(string methodName, int asserts, params string[] assertionMessages)
         {

